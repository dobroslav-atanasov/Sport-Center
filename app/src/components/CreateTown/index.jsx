import React, { Fragment } from 'react';
import MappleToolTip from 'reactjs-mappletooltip';
import validationService from '../../services/validationService';
import townService from '../../services/townService';
import Loading from '../Loading';
import constants from '../../constants/constants';

class CreateTown extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            name: '',
            country: '',
            imageUrl: '',
            errors: {
                name: constants.createTown.TOWN_NAME_REQUIRED,
                country: constants.createTown.COUNTRY_NAME_REQUIRED,
                imageUrl: constants.createTown.IMAGE_URL_REQUIRED
            }
        };
    };

    changeHandler = (event) => {
        event.preventDefault();
        const { name, value } = event.target;
        let errors = this.state.errors;

        switch (name) {
            case 'name':
                if (this.state.towns.includes(value)) {
                    errors.name = constants.createTown.TOWN_EXIST;
                } else if (!validationService.townNameValidation(value)) {
                    errors.name = constants.createTown.INVALID_TOWN;
                } else {
                    errors.name = '';
                }
                break;
            case 'country':
                errors.country = !validationService.countryNameValidation(value) ? constants.createTown.INVALID_COUNTRY : '';
                break;
            case 'imageUrl':
                errors.imageUrl = !validationService.imageUrlValidation(value) ? constants.createTown.INVALID_IMAGE_URL : '';
                break;
            default:
                break;
        };

        this.setState({ errors, [name]: value });
    };

    submitHandler = (event) => {
        event.preventDefault();
        const data = this.state;
        if (validationService.formValidation(this.state.errors)) {
            townService.create(data).then(() => {
                this.props.history.push('/');
            });
        }
    };

    componentDidMount = () => {
        townService.getTowns()
            .then(towns => {
                this.setState({ towns: towns });
            });
    };

    render() {
        const { errors, towns } = this.state;
        return (
            <Fragment>
                {towns ?
                    <div className="container" style={{ marginTop: 30, marginBottom: 50 }}>
                        <div className="card bg-light">
                            <article className="card-body mx-auto">
                                <h3 className="card-title mt-3 text-center">Add Town</h3>
                                <hr />
                                <form onSubmit={this.submitHandler}>
                                    {/* NAME */}
                                    <div className="row">
                                        <div className="col-md-10">
                                            <div className="form-group input-group">
                                                <div className="input-group-prepend">
                                                    <span className="input-group-text"><i className="fa fa-building"></i></span>
                                                </div>
                                                <input id="name" type="text" className="form-control" name="name" onChange={this.changeHandler} placeholder="Town" />
                                            </div>
                                        </div>
                                        <div className="col-md-2">
                                            {errors.name.length === 0
                                                ? <MappleToolTip><div className="text-success"><i className="fa fa-check"></i></div><div className="text-success">Town name is correct!</div></MappleToolTip>
                                                : <MappleToolTip><div className="text-danger"><i className="fa fa-exclamation-triangle"></i></div><div className="text-danger">{errors.name}</div></MappleToolTip>}
                                        </div>
                                    </div>

                                    {/* COUNTRY */}
                                    <div className="row">
                                        <div className="col-md-10">
                                            <div className="form-group input-group">
                                                <div className="input-group-prepend">
                                                    <span className="input-group-text"> <i className="fa fa-globe"></i> </span>
                                                </div>
                                                <input id="country" type="text" className="form-control" name="country" onChange={this.changeHandler} placeholder="Country" />
                                            </div>
                                        </div>
                                        <div className="col-md-2">
                                            {errors.country.length === 0
                                                ? <MappleToolTip><div className="text-success"><i className="fa fa-check"></i></div><div className="text-success">Coutnry is correct!</div></MappleToolTip>
                                                : <MappleToolTip><div className="text-danger"><i className="fa fa-exclamation-triangle"></i></div><div className="text-danger">{errors.country}</div></MappleToolTip>}
                                        </div>
                                    </div>

                                    {/* IMAGE URL */}
                                    <div className="row">
                                        <div className="col-md-10">
                                            <div className="form-group input-group">
                                                <div className="input-group-prepend">
                                                    <span className="input-group-text"> <i class="fa fa-image"></i> </span>
                                                </div>
                                                <input id="imageUrl" type="text" className="form-control" name="imageUrl" onChange={this.changeHandler} placeholder="Image Url" />
                                            </div>
                                        </div>
                                        <div className="col-md-2">
                                            {errors.imageUrl.length === 0
                                                ? <MappleToolTip><div className="text-success"><i className="fa fa-check"></i></div><div className="text-success">Image url is correct!</div></MappleToolTip>
                                                : <MappleToolTip><div className="text-danger"><i className="fa fa-exclamation-triangle"></i></div><div className="text-danger">{errors.imageUrl}</div></MappleToolTip>}
                                        </div>
                                    </div>

                                    {/* SUBMIT */}
                                    <div class="form-group">
                                        <button type="submit" className="btn btn-danger btn-block">Add Town</button>
                                    </div>
                                </form>
                            </article>
                        </div>
                    </div>
                    : <Loading />
                }
            </Fragment>
        );
    };
};

export default CreateTown;