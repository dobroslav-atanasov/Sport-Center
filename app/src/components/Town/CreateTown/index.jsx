import React from 'react';
import MappleToolTip from 'reactjs-mappletooltip';
import validationService from '../../../services/validationService';
import townService from '../../../services/townService';

class Town extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            name: '',
            country: '',
            imageUrl: '',
            towns: [],
            errors: {
                name: 'Town name is required!',
                country: 'Country name is required!',
                imageUrl: 'Image url is required!'
            }
        };
    };

    changeHandler = (event) => {
        event.preventDefault();
        const { name, value } = event.target;
        let errors = this.state.errors;

        switch (name) {
            case 'name':
                if (this.state.towns.includes(value)) {
                    errors.name = 'Town already exist!';
                } else if (!validationService.townNameValidation(value)) {
                    errors.name = 'Town should be at least 3 symbols long and contains only letters!';
                } else {
                    errors.name = '';
                }
                break;
            case 'country':
                errors.country = !validationService.countryNameValidation(value) ? 'Country should be at least 4 symbols long and contains only letters!' : '';
                break;
            case 'imageUrl':
                errors.imageUrl = !validationService.imageUrlValidation(value) ? 'Image url should start with http or https!' : '';
                break;
            default:
                break;
        };

        this.setState({ errors, [name]: value });
    };

    submitHandler = (event) => {
        event.preventDefault();
        const data = this.state;
        if (validationService.formValidation(this.state.errors)) {
            townService.create(data).then(() => {
                this.props.history.push('/');
            });
        }
    };

    componentWillMount = () => {
        townService.getTowns()
            .then(towns => {
                this.setState({ towns: towns });
            });
    };

    render() {
        const { errors } = this.state;
        return (
            <div className="container" style={{ marginTop: 30, marginBottom: 50 }}>
                <div className="card bg-light">
                    <article className="card-body mx-auto">
                        <h3 className="card-title mt-3 text-center">Add Town</h3>
                        <hr />
                        <form onSubmit={this.submitHandler}>
                            {/* NAME */}
                            <div className="row">
                                <div className="col-md-10">
                                    <div className="form-group input-group">
                                        <div className="input-group-prepend">
                                            <span className="input-group-text"><i className="fa fa-building"></i></span>
                                        </div>
                                        <input type="text" className="form-control" name="name" onChange={this.changeHandler} placeholder="Town" />
                                    </div>
                                </div>
                                <div className="col-md-2">
                                    {errors.name.length === 0
                                        ? <MappleToolTip><div className="text-success"><i className="fa fa-check"></i></div><div className="text-success">Town name is correct!</div></MappleToolTip>
                                        : <MappleToolTip><div className="text-danger"><i className="fa fa-exclamation-triangle"></i></div><div className="text-danger">{errors.name}</div></MappleToolTip>}
                                </div>
                            </div>

                            {/* COUNTRY */}
                            <div className="row">
                                <div className="col-md-10">
                                    <div className="form-group input-group">
                                        <div className="input-group-prepend">
                                            <span className="input-group-text"> <i className="fa fa-globe"></i> </span>
                                        </div>
                                        <input type="text" className="form-control" name="country" onChange={this.changeHandler} placeholder="Country" />
                                    </div>
                                </div>
                                <div className="col-md-2">
                                    {errors.country.length === 0
                                        ? <MappleToolTip><div className="text-success"><i className="fa fa-check"></i></div><div className="text-success">Coutnry is correct!</div></MappleToolTip>
                                        : <MappleToolTip><div className="text-danger"><i className="fa fa-exclamation-triangle"></i></div><div className="text-danger">{errors.country}</div></MappleToolTip>}
                                </div>
                            </div>

                            {/* IMAGE URL */}
                            <div className="row">
                                <div className="col-md-10">
                                    <div className="form-group input-group">
                                        <div className="input-group-prepend">
                                            <span className="input-group-text"> <i class="fa fa-image"></i> </span>
                                        </div>
                                        <input type="text" className="form-control" name="imageUrl" onChange={this.changeHandler} placeholder="Image Url" />
                                    </div>
                                </div>
                                <div className="col-md-2">
                                    {errors.imageUrl.length === 0
                                        ? <MappleToolTip><div className="text-success"><i className="fa fa-check"></i></div><div className="text-success">Image url is correct!</div></MappleToolTip>
                                        : <MappleToolTip><div className="text-danger"><i className="fa fa-exclamation-triangle"></i></div><div className="text-danger">{errors.imageUrl}</div></MappleToolTip>}
                                </div>
                            </div>

                            {/* SUBMIT */}
                            <div class="form-group">
                                <button type="submit" className="btn btn-danger btn-block">Add Town</button>
                            </div>
                        </form>
                    </article>
                </div>
            </div>
        );
    };
};

export default Town;