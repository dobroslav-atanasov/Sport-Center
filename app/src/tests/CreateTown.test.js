import React from 'react';
import CreateTown from '../components/CreateTown';
import renderer from 'react-test-renderer';

jest.mock('react-router-dom', () => ({
    Link: "Link"
}));

describe('Create Town component test', () => {
    test('Should render component', () => {
        const component = renderer.create(<CreateTown />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});