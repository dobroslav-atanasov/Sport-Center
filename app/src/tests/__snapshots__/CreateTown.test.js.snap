// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Town component test Should render component 1`] = `
<div
  className="container"
  style={
    Object {
      "height": 100,
      "marginBottom": 50,
      "marginTop": 30,
      "width": 100,
    }
  }
>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\">
  <path opacity=\\".25\\" d=\\"M16 0 A16 16 0 0 0 16 32 A16 16 0 0 0 16 0 M16 4 A12 12 0 0 1 16 28 A12 12 0 0 1 16 4\\"/>
  <path d=\\"M16 0 A16 16 0 0 1 32 16 L28 16 A12 12 0 0 0 16 4z\\">
    <animateTransform attributeName=\\"transform\\" type=\\"rotate\\" from=\\"0 16 16\\" to=\\"360 16 16\\" dur=\\"0.8s\\" repeatCount=\\"indefinite\\" />
  </path>
</svg>
",
      }
    }
    style={
      Object {
        "fill": "#dc3545",
        "height": 64,
        "width": 64,
      }
    }
  />
  <p
    className="text-danger"
  >
    Loading...
  </p>
</div>
`;
